pkg/skaffold/cluster/minikube.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"os/exec"
+
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
+	k8s "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/client"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/context"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+var GetClient = getClient
+
+// To override during tests
+var (
+	minikubeBinaryFunc      = minikubeBinary
+	getRestClientConfigFunc = context.GetRestClientConfig
+)
+
+type Client interface {
+	// IsMinikube returns true if the given kubeContext maps to a minikube cluster
+	IsMinikube(kubeContext string) bool
+	// MinikubeExec returns the Cmd struct to execute minikube with given arguments
+	MinikubeExec(arg ...string) (*exec.Cmd, error)
+}
+
+type clientImpl struct{}
+
+func getClient() Client {
+	return clientImpl{}
+}
+
+func (clientImpl) IsMinikube(kubeContext string) bool {
```

Comment:
what if we cached this in the global config? we're already using kube contexts as keys in the config anyway: we could just add an `isMinikube` field and then check here to short circuit this whole logic.

Reply:
The checks are quite inexpensive now so we might not need a cache. I'm slightly leaning towards not adding this to the global config because:
- it could pollute the config if the user has a workflow where say they use a random named minikube cluster each time and delete that afterwards.
- it could get stale- if they create minikube `test` cluster, delete that and later create kind or gke cluster called `test`
### 🤷🏾‍♂️

Reply:
Maybe we should have a KubeContext type instead of passing around string.

Reply:
created https://github.com/GoogleContainerTools/skaffold/issues/4751